Sources/Leaf/Parser.swift

```diff
@@ -0,0 +1,347 @@
+            if byte == .leftCurlyBracket {
+                body = try extractBody()
+            } else {
+                body = nil
+            }
+        } else {
+            body = nil
+        }
+
+        let kind: SyntaxKind = .tag(
+            name: id,
+            parameters: params,
+            body: body
+        )
+        
+        let source = Source(line: line, column: column, range: start..<scanner.offset)
+        return Syntax(kind: kind, source: source)
+    }
+
+    private func extractBody() throws -> [Syntax] {
+        try expect(.leftCurlyBracket)
+        let body = try bytes(until: .rightCurlyBracket)
+        try expect(.rightCurlyBracket)
+        let parser = Parser(body)
+        return try parser.parse()
+    }
+
+    private func extractRaw() throws -> Bytes {
+        return try bytes(until: .numberSign)
+    }
+
+    private func bytes(until: Byte) throws -> Bytes {
+        var previous: Byte?
+
+        var bytes: Bytes = []
+        while let byte = scanner.peek(), byte != until || previous == .backSlash {
+            try scanner.pop()
+            if byte != until && previous == .backSlash {
+                bytes.append(.backSlash)
+            }
+            if byte != .backSlash {
+                bytes.append(byte)
+            }
+            previous = byte
+        }
+
+        return bytes
+    }
+
+    private func extractIdentifier() throws -> Syntax {
+        let start = scanner.offset
+        let line = scanner.line
+        let column = scanner.column
+
+        while let byte = scanner.peek(), byte.isAlphanumeric {
+           try scanner.pop()
+        }
+        
+        let bytes = scanner.bytes[start..<scanner.offset]
+
+        let kind: SyntaxKind = .identifier(name: bytes.makeString())
+        let source = Source(line: line, column: column, range: start..<scanner.offset)
+
+        return Syntax(kind: kind, source: source)
+    }
+
+    private func extractParameters() throws -> [Syntax] {
+        try expect(.leftParenthesis)
+
+        var params: [Syntax] = []
+        repeat {
+            if params.count > 0 {
+                try expect(.comma)
+            }
+
+            if let param = try extractParameter() {
+                params.append(param)
+            }
+        } while scanner.peek() == .comma
+
+        try expect(.rightParenthesis)
+
+        return params
+    }
+
+    private func extractNumber() throws -> Constant {
+        let start = scanner.offset
+        while let byte = scanner.peek(), byte.isDigit || byte == .period {
+            try scanner.pop()
+        }
+
+        let bytes = scanner.bytes[start..<scanner.offset]
+        let string = bytes.makeString()
+        if bytes.contains(.period) {
+            guard let double = Double(string) else {
+                throw "Unexpected non double"
+            }
+            return .double(double)
+        } else {
+            guard let int = Int(string) else {
```

Comment:
Note: Since your parser doesn't support negatives, you're really expecting `UInt64` or you'll overflow and get a poison value/runtime crash.